### DISCUSSION TITLE
View Binding에서 binding.root의 정확한 의미

### DISCUSSION THREAD

**Discussion by JuTaK97** [wafflestudio/seminar-2023#162] [Q&A] [answered]
[원본 질문(슬랙)](https://wafflestudio2-k6c7167.slack.com/archives/C05NC5R1VSA/p1696086000463869)
```
과제 수행 전, 영상으로 복습하던 중 질문이 생겨 글 올립니다. root 관련 질문입니다.
setContentView(Binding.root)를 onCreate 함수에 넣는 이유가 무엇인가요? R.layout.MainActivity는 안되는건가요? 
Binding.root가 의미하는게 무엇인가요?
새로운 layout을 만들 때,  root를 설정하는 이유가 무엇이고, 그 root라는 것의 역할이나 의미가 무엇인가요?
```
좋은 질문 by @kkomul1  🧇🧇🧇

**Comment by JuTaK97** [comment, marked as answer]
## 1. setContentView와 xml inflate에 대해서

액티비티 클래스는 setContentView 함수에 대해 여러 overload를 가지고 있습니다. 
view binding을 사용하기 전에는 resource의 id (`R.layout.activity_main`) 을 인자로 받는 setContentView를 사용했고, view binding을 사용한 후에는 `View`를 직접 인자로 받는 다른 setContentView를 사용합니다.

> 첫 번째 setContentView
```
    @Override
    public void setContentView(int resId) {
        ensureSubDecor();
        ViewGroup contentParent = mSubDecor.findViewById(android.R.id.content);
        contentParent.removeAllViews();
        LayoutInflater.from(mContext).inflate(resId, contentParent);
        mAppCompatWindowCallback.bypassOnContentChanged(mWindow.getCallback());
    }
```

> 두 번째 setContentView
```
    @Override
    public void setContentView(View v) {
        ensureSubDecor();
        ViewGroup contentParent = mSubDecor.findViewById(android.R.id.content);
        contentParent.removeAllViews();
        contentParent.addView(v);
        mAppCompatWindowCallback.bypassOnContentChanged(mWindow.getCallback());
    }
```

`R.layout.activity_main` 은 xml이라는 데이터 표현 형식을 빌려서 만든 뷰의 설계도 파일에 불과하고, 우리는 이것을 실제 `View` 객체로 바꿔 줘야 합니다. 이 과정이 inflate입니다. 
두 setContentView를 비교해 보면, resId를 받아서 inflate를 시키는 과정만 추가된 것을 볼 수 있습니다. 더 깊이 들어가서 코드를 읽는 것도 좋지만 당장은 이해가 쉽지 않을 것입니다. (저도 어렵습니다)

아무튼, setContentView는 액티비티(window) 내의 최상위 View를 추가하는 함수입니다. 액티비티의 생명주기에 따르면, onCreate는 액티비티의 생애 주기에서 단 한번 불리는 함수입니다. onCreate에서는 각종 UI나 리스너의 세팅, 데이터 바인딩 등을 수행하는 것을 권장하고, setContentView를 onCreate에서 수행합니다.
참고 : https://stackoverflow.com/questions/33973399/why-do-we-set-content-view-in-oncreate-and-not-in-onstart-or-onresume


## binding.root 에 대해서

view binding을 사용하면, [ViewBinding](https://cs.android.com/androidx/platform/frameworks/data-binding/+/mirror-goog-studio-master-dev:extensions/viewbinding/src/main/java/androidx/viewbinding/ViewBinding.java?q=viewbinding)을 extend하는 클래스가 각 xml 파일에 대응되어 자동으로 생성됩니다. (`activity_main.xml` -> `ActivityMainBinding`)
inflate를 통해 ActivityMainViewBinding 클래스의 인스턴스를 만들고, 보통 `binding`이라고 명명합니다. 그리고 `binding.root`, 즉 `binding.getRoot()` 를 하면 View를 가져올 수 있습니다.
이때 얻는 root는 해당 xml 파일로 표현한 가장 바깥쪽 계층의 View입니다. 이번 과제를 예로 들면

```
<!-- activity_main.xml -->
<?xml version="1.0" encoding="utf-8"?>
<androidx.drawerlayout.widget.DrawerLayout xmlns:android="http://schemas.android.com/apk/res/android"
 xmlns:app="http://schemas.android.com/apk/res-auto"
 xmlns:tools="http://schemas.android.com/tools"
 android:layout_width="match_parent"
 android:layout_height="match_parent"
 tools:context=".MainActivity">

<!-- 이 부분에는 메인 화면의 UI를 넣습니다. -->

 <com.google.android.material.navigation.NavigationView
     android:id="@+id/drawer"
     android:layout_width="match_parent"
     android:layout_height="match_parent">

    <!-- 이 부분에는 서랍 안 화면의 UI를 넣습니다. -->

 </com.google.android.material.navigation.NavigationView>

</androidx.drawerlayout.widget.DrawerLayout>
```
의 경우, 해당 xml파일을 inflate해서 View Binding을 만들어서 `getRoot()`를 하면 가장 바깥쪽의 DrawerLayout 객체 (View) 를 얻게 됩니다.

<img width="533" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/184fbff6-52ce-449d-8663-d539d5bb7b0b">

<img width="536" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/a59f8d82-56a1-4ccc-ae2f-6762b75e4baf">

위와 같이 타입 추론까지 해 주네요 (코틀린 짱)

정리하자면, root를 "우리가 설정하는" 것이 아니라, view binding에 의해 `main_activity.xml`에 대응하는 `ActivityMainBinding` 클래스가 자동으로 만들어지고, inflate를 통해 해당 클래스의 인스턴스(binding)을 얻으면 root가 자연스럽게 생기는 것입니다. 우리는 ViewBinding 인터페이스에서 제공하는 `getRoot` 메서드를 통해 xml 파일 상 가장 바깥쪽 계층의 View를 가져와서 사용하기만 하면 됩니다.

참고로 ViewBinding이 아니라 그냥 View에서도 해당 View가 속한 ViewGroup 상으로 가장 바깥쪽 계층의 View를 `.getRootView()` 를 통해 얻을 수 있습니다. 아래와 같이 로그를 찍어 보면, root 입장에서 root는 자기 자신임을 알 수 있습니다.
<img width="1012" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/98825412-3cf7-4845-8b39-c137f6f03889">

<img width="979" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/21af812c-8141-4056-a259-f28a88f26234">
레이아웃 상 DrawerLayout의 child인 NavigationView한테 rootView를 달라 하면 DrawerLayout을 가져다 주는 것도 볼 수 있네요

**Reply by kkomul1** [reply]
추가로 더 질문이 있어서 남깁니다.
1. ActivityMainBinding class의 method인 inflate()는 무엇을 함으로써 ActivityMainBinding class의 instance를 반환하는건가요? 생성자 형식으로 되어있는게 아닌 것 같아서 질문합니다. Ctrl + b로 찾아가 봐도 제가 만든 xml 파일만 나오네요.(이런건 뜯어보려먼 어떻게 해야하나요? -> 8번 질문입니다)
2. 또한, inflate()가 받고 있는 매개변수인 layoutInflater는 어떤 녀석인지 궁금합니다. 추측으로는 그냥 말 그대로 xml을 view로 바꿔주는 함수인 것 같은데, inflate()라는 함수 안에 또 inflate에 관련된 것을 넣는게 좀 이상해서요.
3. binding.root와 binding.getRoot()의 반환값이 같다면 왜 굳이 두 가지 방법을 만들어 뒀을까요? root를 ViewBinding class의 public field로 선언해두고 그걸 그대로 상속받아서 쓰도록 만들어 두었으리라 추측되는데, 왜 굳이 root 자체에 접근할 수 있도록 해두었을까요?(root를 직접 수정하면 조금 골치 아파 질 것 같은데..)
4. 'binding.root에 대해서'의 1문단 3줄에서 binding이 ActivityMainViewBinding class의 instance라고 하셨는데, 제가 이해한 바로는 ActivityMainBinding class의 instance가 옳은 것 같습니다.
5. setContentView 함수에 넘겨주는 값이 binding.root가 아닌 R.layout.activity_main을 사용해도 상관 없어야 하는 것 아닌가요? 결국 binding.root는 activity_main의 최상위 view를 가리키게 되는데, R.layout.activity_main을 사용하면 setContentView가 이를 알아서 inflate해서 activity_main의 최상위 view로 바꾸는 것으로 이해하였습니다. 이렇게 되면 두 가지 방법이 차이가 없어야 하는데, 꼭 binding.root로 바꿔주어야 하는 이유가 궁금합니다. (xml을 한 번 inflate하는 cost를 줄일 수는 있겠죠)
6. 제 추측으로는 setContentView() 함수로 Activity가 시작될 때(onCreate() 함수가 실행될 때) 최상위 view를 화면에 띄운 뒤, 그 위에 쌓이는 다른 view들은 최상위 view에 대한 상대적 위치와 크기로 정의되어 있으므로 따로 setContentView()를 통해 화면에 띄울 필요가 없는 것 같습니다. 제 추측이 올바른가요?
7. chatGPT로 ViewBinding class의 내부 구현을 살짝 들여다 봤는데, 여기서도 getViewById()를 이용하여 view를 가져오는 것을 볼 수 있었습니다. 그렇다면 viewBinding을 이용한 성능 향상이 딱히 없는 것 같은데, ppt 8쪽에서는 레이아웃 계층이 커질 때의 성능 저하를 방지하기 위해 사용한다고 쓰여 있습니다. 어떤 의미에서 하신 말씀일까요?
8. 저런 함수들이나 객체들(inflate, binding, binding.root 등)에 커서를 두고 Ctrl+b를 누르면 definition이 나오는게 아니라 그 함수나 객체가 지칭하는 녀석이 나오게 됩니다. 그 이유가 무엇이고, definition에 접근하려면 어떻게 해야 할까요?

**Reply by JuTaK97** [reply]
하나씩 답변드릴게요

1 & 8. ViewBinding의 구현부들은 `app/build/generated/data_binding_base_class_source_out/debug/out/com/~~/~~/databinding` 에서 보실 수있습니다. 
ctrl+b로 바로 연결되는 건 아마 컴파일러 수준에서 generated된 코드라 그렇게 연결될 수 있는 걸로 알고 있습니다

**Reply by JuTaK97** [reply]
2. 1번에서 찾은 source의 코드를 보면 이해가 되실 것 같은데요,
```
  @NonNull
  public static ActivityMainBinding inflate(@NonNull LayoutInflater inflater,
      @Nullable ViewGroup parent, boolean attachToParent) {
    View root = inflater.inflate(R.layout.activity_main, parent, false);
    if (attachToParent) {
      parent.addView(root);
    }
    return bind(root);
  }
```
결국 내부에서 layoutinflator.inflate()를 부르네요.

LayoutInflater는 xml을 view로 바꿔주는 역할을 하는 게 맞습니다. 찾아보면 [이런](https://www.youtube.com/watch?v=Y06wmVIFlsw) 것도 있긴 한데... Android의 View나 Context에 관한 깊고 심오한 지식이 있어야 구현부를 이해할 수 있을 거라 저는 더 레벨업 하고 공부해 보겠습니다.

**Reply by JuTaK97** [reply]
3. root에 대한 getter만 제공하는 거라 맘대로 수정할 염려는 없어보입니다.(코드를 보니 final이기도 하네요) 그리고 binding.root와 view.rootView는 다릅니다! binding은 ActivityMainBinding의 인스턴스입니다.
4. 맞네요 오타입니다.
5. 두 가지 모두 당연히 옳은 방법입니다. view binding을 쓰냐 마냐의 차이인데, view binding을 사용할 때는 setContentView(binding.root)로 하라고 [공식 가이드](https://developer.android.com/topic/libraries/view-binding)에서 안내하고 있기 때문에 모두가 이 방법을 따르는 것입니다.

**Reply by JuTaK97** [reply]
6. setContentView로 전달하는 최상위 뷰 자체가 내부에 계층 구조를 갖는 view입니다. 말씀하신대로 child view들은 parent에 대한 상대적 위치, 그리고  자신의 param(match_parent, wrap_content 등) 등에 따라서 위치가 정해지고 그려지게 됩니다. 이 과정에 대한 자세한 설명은 [공식 가이드](https://developer.android.com/guide/topics/ui/how-android-draws) 를 포함한 다양한 레퍼런스에서 찾아볼 수 있습니다.

**Reply by JuTaK97** [reply]
7. 정말 모범적으로, 한 child에 대해 딱 한번씩만 findViewById를 하면 결국 하는 건 똑같으므로 속도는 비슷할 것 같습니다(view binder 컴파일러가 최적화한 로우레벨에서의 뭔가 차이는 있을지도 모르겠네요). 하지만 findViewById는 순회해서 찾고 끝이지만, view binding은 inflate 시점에서 모두 다 참조 가능한 객체로 저장하기 때문에 여러 번 찾게 될 경우 효율적입니다. (모범적으로 하려면 모든 child view를 Activity의 public property로 lateinit으로 다 만들어서 onCreate에서 하나하나 findByViewId를 해야 되는데, 매우 보기 좋지 않죠)

성능은 부차적인 이점이고, null-safety나 가독성이 view binding의 주요한 이점이라고 보면 좋을 것 같네요.

**Reply by kkomul1** [reply]
하나하나 자세히 답변해주셔서 감사합니다! 그런데 3번 답변에서 binding.root와 view.rootView가 어떤 차이가 있는지 여쭤봐도 될까요? 둘 다 액티비티의 최상위 뷰를 가리키는 것 같은데 어떤 차이를 말씀하시는 것인지 모르겠습니다.

**Reply by JuTaK97** [reply]
binding.root는 ViewBinding 객체에서 root를 가져오는 것입니다. (xml 파일 상 최상의 view)

view.rootView는 View 객체에서 그 View가 속한 계층 상의 root를 가져오는 것입니다.

binding.root 가 뷰 계층 내에서 최상위라서 binding.root.rootView 와 binding.root 가 같은 객체라고 말씀드렸던 것이고, view.rootView는 binding.root와는 다른 함수입니다.

**Reply by kkomul1** [reply]
그러면 하나의 xml 파일에서 ConstraintLayout 안에 TextView1과 LinearLayout이 있고 그 LinearLayout 안에 TextView2가 있으면, 그 xml 파일을 inflate 해서 만들어진 binding에 대하여 binding.root, binding.root.rootView, TextView1.rootView, 그리고 TextView2.rootView는 모두 xml 파일의 ConstraintLayout을 가리키는 건가요?

**Reply by JuTaK97** [reply]
너무 좋은 질문입니다. 답은 "상황에 따라 다르다" 입니다.

xml 파일을 inflate한 binding을 생성한 직후에는 binding.root 는 스스로가 root입니다. 아직 해당 View가 어디에도 붙지 않았고 그냥 단순히 inflate만 시킨 상태이기 때문입니다.
하지만 setContentView(binding.root) 를 하는 순간, xml을 inflate해서 만든 이 View는 Activity에 붙습니다. 그리고 프레임워크 단에서 Activity 의 최상위 view는 [decorView](https://stackoverflow.com/questions/23276847/what-is-an-android-decorview)라고 불리는 View이이기 때문에, setContentView 이후에는 view들의 rootView가 decorView가 됩니다. 

실제 로그를 찍어보겠습니다. 과제1의 레이아웃에서 setContent 전후로 binding 하위 여러 뷰들, xml 파일 상 최상위 뷰인 binding.drawerLayout, 그리고 binding.root과 decorView의 hashcode를 찍어 보았습니다.
```kotlin
super.onCreate(savedInstanceState)
binding = ActivityMainBinding.inflate(layoutInflater)
Log.d("aaaa", "decorView : ${window.decorView.hashCode()}")
Log.d("aaaa", "drawerLayout과 binding.root :${binding.drawerLayout.hashCode()} ${binding.root.hashCode()}")

Log.d("aaaa", "setContentView 전 view들의 rootView : ${binding.linear1.rootView.hashCode()} ${binding.const1.rootView.hashCode()} ${binding.cell1.rootView.hashCode()}\n")
Log.d("aaaa", "setContentView 전 binding.root의 rootView : ${binding.root.rootView.hashCode()}")
setContentView(binding.root)
Log.d("aaaa", "setContentView 후 view들의 rootView : ${binding.linear1.rootView.hashCode()} ${binding.const1.rootView.hashCode()} ${binding.cell1.rootView.hashCode()}")
Log.d("aaaa", "setContentView 후 binding.root의 rootView : ${binding.root.rootView.hashCode()}")
```
<img width="600" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/6c57d3dd-716e-4842-9066-494104cc28c6">

이렇게 로그를 찍어 보면 setContentView 전에는 binding에 있는 뷰들은 모두 rootView가 최상단 drawerLayout으로 동일합니다.
하지만 setContentView를 하면 뷰들의 rootView는 다른 객체로 바뀌고, 이는 액티비티에서 `this.window.decorView` 로 접근할 수 있는 decorView임을 알 수 있습니다.

**Comment by JuTaK97** [comment]
### 전반적인 조언

저희는 공부해야 될 게 무척 많습니다. 프레임워크의 내부 구현도 궁금하긴 하지만 지금 공부하기에는 너무 복잡하고, 너무 어렵고, 사전 지식이 무척 많아야 합니다. 지금은 기능 구현이나 인터페이스의 의미와 사용법을 익히는 데 집중하고, 실제 구현부는 충분히 프레임워크 사용에 익숙해 졌을 때(~짬이 찼을때~) 차차 공부해 가는 것을 권장드립니다. 물론 공식 가이드에 있는 내용들은 지금 공부하시면 아주 좋습니다! 

만약 공식 가이드에서 일정 지점부터 코드를 보지 않는 한 이해할 수 없고 설명이 부족하다면, 저는 그것은 알 필요가 없거나 가이드를 읽는 레벨에서는 알 수 없는 내용이라고 생각하고 넘어갑니다. 한 주제에 대해서 깊은 관심을 가지고 소스 코드를 한줄한줄 이해하며 deep dive해 보는 것도 정말 훌륭하고 멋진 일이라고 생각하지만, 저는 현재 단계에서 공부할 수 있는 다른 것들을 공부하는 게 더 중요하다고 생각합니다.

**Reply by kkomul1** [reply]
네 알겠습니다! 깊게 파는건 어느정도까지만 하고 일단 구현부터 열심히 해서 과제 제출해보도록 할게요~ :)
